agent/websockets/shim.go

```diff
@@ -307,24 +312,26 @@ func createShimChannel(ctx context.Context, host, shimPath string) http.Handler
 			http.Error(w, fmt.Sprintf("internal error reading a shim request: %v", err), http.StatusInternalServerError)
 			return
 		}
-		var msg sessionMessage
-		if err := json.Unmarshal(body, &msg); err != nil {
+		var msgs []sessionMessage
+		if err := json.Unmarshal(body, &msgs); err != nil {
 			http.Error(w, fmt.Sprintf("error parsing a shim request: %v", err), http.StatusBadRequest)
 			return
 		}
-		c, ok := connections.Load(msg.ID)
-		if !ok {
-			http.Error(w, fmt.Sprintf("unknown shim session ID: %q", msg.ID), http.StatusBadRequest)
-			return
-		}
-		conn, ok := c.(*Connection)
-		if !ok {
-			http.Error(w, "internal error reading a shim session", http.StatusInternalServerError)
-			return
-		}
-		if err := conn.SendClientMessage(msg.Message); err != nil {
-			http.Error(w, fmt.Sprintf("attempt to send data on a closed session: %q", msg.ID), http.StatusBadRequest)
-			return
+		for _, msg := range msgs {
+			c, ok := connections.Load(msg.ID)
```

Comment:
This seems like a change in behavior.

Before, if one websocket command failed, the subsequent commands would still pass.
Now, if one command fails, all the subsequent commands in the same batch now get skipped

Is there some mechanism that I'm not noticing which ensure the other batched commands still get executed?

Reply:
The code could probably be restructured to be clearer.

The behavior is unchanged. The only way the called method can return an error is if the connection has been closed, which is a terminal state.

That means that in the previous logic, if one send failed then all of the subsequent ones would also fail.

... as such, this change doesn't modify the behavior for subsequent requests. Both before and after this change those would have failed.